llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp

```diff
@@ -0,0 +1,186 @@
+      continue;
+
+    // Get a pointer to the entry point function from the metadata.
+    const MDOperand &FuncOperand = MD->getOperand(0);
+    if (!FuncOperand)
+      continue;
+    if (auto *FuncConstant = dyn_cast<ConstantAsMetadata>(FuncOperand))
+      if (auto *Func = dyn_cast<Function>(FuncConstant->getValue()))
+        if (Func->getName() == KernelName)
+          return MD;
+  }
+
+  return nullptr;
+}
+
+PreservedAnalyses CUDANativeSpecConstantsPass::run(Module &M,
+                                                   ModuleAnalysisManager &MAM) {
+  NamedMDNode *MD = M.getNamedMetadata(::KERNEL_SPEC_CONST_MD_STRING);
+  if (!MD)
+    return PreservedAnalyses::all();
+
+  // NOTE: GlobalNamePrefix is read by pi_cuda and must not be changed.
+  constexpr char GlobalNamePrefix[] = "sycl_specialization_constants_kernel_";
+  DenseMap<StringRef, GlobalVariable *> Globals;
+
+  // Calculate the size of global symbol and allocate it for each kernel.
+  for (const auto *Node : MD->operands()) {
+    const StringRef KernelName =
+        cast<MDString>(Node->getOperand(0).get())->getString();
+    unsigned PerKernelSize = 0;
+    // Loop over all spec constants of a kernel
+    for (unsigned i = 1; i < Node->getNumOperands(); ++i) {
+      MDNode *SC = dyn_cast<MDNode>(Node->getOperand(i));
+      assert(SC->getNumOperands() >= 4 && "Invalid node.");
+      // get the size and offset node to calculate the total size of spec
+      // constants (size of type + offset in the composite - if any);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 1);
+      PerKernelSize += uintFromMDNode(SC, SC->getNumOperands() - 2);
+    }
+    // finally, allocate per-kernel globals.
+    std::string GlobalName{GlobalNamePrefix};
+    GlobalName.append(KernelName.str());
+    Globals[KernelName] = createGlobal(M, PerKernelSize, GlobalName);
+  }
+
+  auto &Context = M.getContext();
+  IRBuilder B(Context);
+  SmallVector<Instruction *> ToErase;
+  // Now, re-write the signature of the kernel to drop the extra spec constant
+  // argument and patch in accesses to the global symbol (instead of the arg).
+  for (const auto *Node : MD->operands()) {
+    const StringRef KernelName =
+        cast<MDString>(Node->getOperand(0).get())->getString();
+    auto *F = M.getFunction(KernelName);
+    assert(F && "Function not found.");
+    auto *SpecConstArg = std::prev(F->arg_end());
+    assert(SpecConstArg->getName() == "_arg__specialization_constants_buffer" &&
+           "Expected a kernel.");
+
+    // Prepare a new function type, a copy of the original without the last arg.
+    std::vector<Type *> Arguments;
+    SmallVector<AttributeSet, 8> ArgumentAttributes;
+    auto FAttrs =
+        F->getAttributes().removeParamAttributes(Context, F->arg_size() - 1);
+    for (unsigned i = 0; i < F->arg_size() - 1; ++i) {
+      Arguments.push_back(F->getArg(i)->getType());
+      ArgumentAttributes.push_back(FAttrs.getParamAttrs(i));
```

Comment:
```suggestion
    for (unsigned I = 0; I < F->arg_size() - 1; ++I) {
      Arguments.push_back(F->getArg(I)->getType());
      ArgumentAttributes.push_back(FAttrs.getParamAttrs(I));
```

Reply:
:) that's the clang tidy warning that I have off, will update